sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4901,6 +4925,18 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
   PI_ASSERT(Event, PI_INVALID_EVENT);
   PI_ASSERT((WorkDim > 0) && (WorkDim < 4), PI_INVALID_WORK_DIMENSION);
 
+  // Get a new command list to be used on this call
+  pi_command_list_ptr_t CommandList{};
+  {
+    std::scoped_lock QueueLock(Queue->Mutex);
+    if (auto Res = Queue->Context->getAvailableCommandList(
+            Queue, CommandList, false /* UseCopyEngine */,
+            true /* AllowBatching */))
+      return Res;
+  }
+
+  // Lock automatically releases when this goes out of scope.
+  std::scoped_lock Lock(Queue->Mutex, Kernel->Mutex);
```

Comment:
is it OK that the "Queue" can be operated by another thread after a suitable command-list is found above? why don't we hold the queue lock for the duration of this entire function?

Reply:
Fixed this. After the change https://github.com/intel/llvm/pull/6168 we can lock queue and kernel simultaneously for the scope of this function.